llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-simplify-image-buffer-stores.ll

```diff
@@ -23,7 +23,8 @@ define amdgpu_ps void @image_store_1d_store_insert_zeros_at_end(<8 x i32> inreg
 ; GCN-NEXT:    ret void
 ;
 ; GFX12-LABEL: @image_store_1d_store_insert_zeros_at_end(
-; GFX12-NEXT:    call void @llvm.amdgcn.image.store.1d.f32.i32(float [[VDATA1:%.*]], i32 1, i32 [[S:%.*]], <8 x i32> [[RSRC:%.*]], i32 0, i32 0)
+; GFX12-NEXT:    [[NEWVDATA4:%.*]] = insertelement <4 x float> <float poison, float 0.000000e+00, float 0.000000e+00, float 0.000000e+00>, float [[VDATA1:%.*]], i64 0
+; GFX12-NEXT:    call void @llvm.amdgcn.image.store.1d.v4f32.i32(<4 x float> [[NEWVDATA4]], i32 15, i32 [[S:%.*]], <8 x i32> [[RSRC:%.*]], i32 0, i32 0)
```

Comment:
This resulted in using a larger register type, which is worse?

Reply:
Right. The behaviour changed in GFX12, so the old trailing zeros optimization no longer applies (which is what this diff shows) but the new broadcast optimization applies instead.

Reply:
Seems like an unfortunate change. Is the old optimization *wrong*? I have a lot more questions and concerns if so. We can't let optimizations run on IR without a set target-cpu break a program that ultimately executes on gfx12. Do we need an optimization that unfolds the scalar form to the vector with broadcast?

Reply:
No the old optimization is not wrong. GFX12 has an incompatible change in the behaviour of the instruction. The old optimization takes advantage of the pre-GFX12 behaviour, and the new optimization takes advantage of the GFX12 behaviour. This relies on the subtarget being set correctly so we know whether we're compiling for GFX12 or not - is that a problem?

Reply:
> We can't let optimizations run on IR without a set target-cpu break a program that ultimately executes on gfx12.

So is it OK for this optimization to do different things for GFX12 and pre-GFX12? Do we need a third option where it does nothing if the target CPU is not known?

Reply:
So we need to definitively know if we're gfx12 or not. Ideally we would have a has-gfx12-behavior and has-pre-gfx12-behavior subtarget feature, and the combine would check both